cpp/src/interop/from_arrow_host.cu

```diff
@@ -0,0 +1,445 @@
+    if (bitmap == nullptr) { return std::make_unique<rmm::device_buffer>(0, stream, mr); }
+
+    auto const bitmask_size = array->length + array->offset;
+    auto const allocation_size =
+      bitmask_allocation_size_bytes(static_cast<size_type>(bitmask_size));
+    auto mask = std::make_unique<rmm::device_buffer>(allocation_size, stream, mr);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(mask->data(),
+                                  reinterpret_cast<uint8_t const*>(bitmap),
+                                  allocation_size,
+                                  cudaMemcpyDefault,
+                                  stream.value()));
+    return mask;
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(not is_rep_layout_compatible<T>() &&
+                           !std::is_same_v<T, numeric::decimal128>)>
+  std::unique_ptr<column> operator()(ArrowSchemaView*, ArrowArray const*, data_type, bool)
+  {
+    CUDF_FAIL("Unsupported type in copy_from_arrow_device.");
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(is_rep_layout_compatible<T>() || std::is_same_v<T, numeric::decimal128>)>
+  std::unique_ptr<column> operator()(ArrowSchemaView* schema,
+                                     ArrowArray const* input,
+                                     data_type type,
+                                     bool skip_mask)
+  {
+    using DeviceType = std::conditional_t<std::is_same_v<T, numeric::decimal128>, __int128_t, T>;
+
+    size_type const num_rows   = input->length;
+    size_type const offset     = input->offset;
+    size_type const null_count = input->null_count;
+    auto data_buffer           = input->buffers[fixed_width_data_buffer_idx];
+
+    auto const has_nulls = skip_mask ? false : input->buffers[validity_buffer_idx] != nullptr;
+    auto col = make_fixed_width_column(type, num_rows, mask_state::UNALLOCATED, stream, mr);
+    auto mutable_column_view = col->mutable_view();
+    CUDF_CUDA_TRY(
+      cudaMemcpyAsync(mutable_column_view.data<DeviceType>(),
+                      reinterpret_cast<uint8_t const*>(data_buffer) + offset * sizeof(DeviceType),
+                      sizeof(DeviceType) * num_rows,
+                      cudaMemcpyDefault,
+                      stream.value()));
+
+    if (has_nulls) {
+      auto tmp_mask = get_mask_buffer(input);
+
+      // if array is sliced, we have to copy the whole mask and then take copy
+      auto out_mask =
+        (offset == 0)
+          ? std::move(*tmp_mask)
+          : cudf::detail::copy_bitmask(
+              static_cast<bitmask_type*>(tmp_mask->data()), offset, offset + num_rows, stream, mr);
+
+      col->set_null_mask(std::move(out_mask), null_count);
+    }
+
+    return col;
+  }
+};
+
+// forward declaration is needed because `type_dispatch` instantiates the
+// dispatch_copy_from_arrow_host struct causing a recursive situation for struct,
+// dictionary and list_view types.
+std::unique_ptr<column> get_column_copy(ArrowSchemaView* schema,
```

Comment:
Could we also have a brief comment on when this function is needed? It's not immediately obvious without inspecting all the call sites since it's basically a wrapper around the dispatch functor with some special handling for empty columns